src/PluginStartupRegistration.cpp

```diff
@@ -100,31 +208,68 @@ const std::vector<wxString>& PluginStartupRegistration::GetFailedPluginsPaths()
    return mFailedPluginsPaths;
 }
 
-void PluginStartupRegistration::Run()
+void PluginStartupRegistration::Run(std::chrono::seconds timeout)
 {
-   auto dialog = BasicUI::MakeProgress(XO("Searching for plugins"), XO(""));
-   ProcessNext();
-   while(mValidator)
+   PluginScanDialog dialog(nullptr, wxID_ANY, XO("Searching for plugins"));
+   wxTimer timeoutTimer(&dialog, OnPluginScanTimeout);
+   mScanDialog = &dialog;
+   mTimeoutTimer = &timeoutTimer;
+   mTimeout = timeout;
+
+   auto cleanup = finally([this]
    {
-      const auto message = TranslatableString { mPluginsToProcess[mCurrentPluginIndex].first, { } };
-      //Update UI
-      if(dialog->Poll(mCurrentPluginIndex, mPluginsToProcess.size(), message) != BasicUI::ProgressResult::Success)
-         Stop();
-      //AsyncPluginValidator uses event loop for internal message
-      //delivery, but ProgressDialog::Poll implementation does not call
-      //wxApp::Yield each time, which may result in too long CPU stalls
-      if(wxTheApp->HasPendingEvents())
-         wxTheApp->Yield();
+      mTimeoutTimer = nullptr;
+      mScanDialog = nullptr;
+   });
+
+   dialog.Bind(wxEVT_BUTTON, [this](wxCommandEvent& evt) {
+      evt.Skip();
+      if(evt.GetId() == wxID_IGNORE)
+         Skip();
+   });
+   dialog.Bind(wxEVT_TIMER, [this](wxTimerEvent& evt) {
+      if(evt.GetId() == OnPluginScanTimeout)
+      {
+         if(mValidator->InactiveSince() < mRequestStartTime)
+            Skip();
+         //else
+         //   wxMessageBox("Please check for plugin popups!");
```

Comment:
What's the activity check really doing if there is no else branch?  But I'm still reviewing the second commit, so let me see.

Reply:
I understand better now, I see where mRequestStartTime is assigned.